File: sdks/python/tests/test_optimistic.py
"""The cursor-gated optimistic-layer engine, against the shared golden scenarios.

Every expectation is read from ``protocol/fixtures/offline-optimistic.json`` so
this port and the other six assert the same values rather than each documenting
its own behaviour.

Server frames are fed to the REAL ``LunoraClient.handle_frame``. This suite used
to fold them through a hand-copied transcription of the client's ``data`` branch,
which meant every one of these cases passed whatever the production handler did —
and a cursor bug in exactly that handler is what the ``cursorlessFrame`` case
below exists for. The client needs no network to be driven: the sender and the
HTTP poster are both injectable.
"""

from __future__ import annotations

import asyncio
import os
import sys
import unittest

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from lunora.client import LunoraClient
from lunora.optimistic import OptimisticLayer, OptimisticLocalStore, apply_optimistic_layer
from lunora.submit import SubmitOptions
from tests._fixtures import load
from tests._manifest import covers

FIXTURES = load("offline-optimistic.json")["optimistic"]

PATH = "messages:list"
ARGS = {"channel": "general"}


def _appender(item):
    """The one transform primitive the fixtures use: push onto a COPY of the list.

    A copy, not an in-place append: a transform is re-run on every rebase, so one
    that mutated its input would compound its own effect on each server frame.
    """

    return lambda current: [*(current or []), item]


def _run(deferred):
    for call in deferred:
        call()


def _live(base, cursor=None):
    """A real client with one subscription, primed with a server value.

    Returns ``(client, subscription, seen)``. The priming frame goes through
    ``handle_frame`` like any other, so the subscription's cursor is whatever the
    production handler decided it should be.
    """

    client = LunoraClient("https://app.example")
    seen: list = []
    client.attach_socket(lambda _frame: None)
    client.subscribe(PATH, ARGS, seen.append)
    _frame(client, {"data": base} if cursor is None else {"cursor": cursor, "data": base})

    return client, client._subs["sub_1"], seen


def _frame(client, frame):
    """Feed one ``data`` frame for ``sub_1`` to the client's real dispatcher."""

    return client.handle_frame({**frame, "id": "sub_1", "type": "data"})


class TestOptimisticRebase(unittest.TestCase):
    def test_layer_rebases_onto_a_later_server_frame(self):
        covers("optimistic_layer_rebases_onto_server_frame")
        case = FIXTURES["rebase"]
        client, sub, seen = _live(case["base"])

        deferred = []
        apply_optimistic_layer(sub, _appender(case["appended"]), deferred)
        _run(deferred)

        self.assertEqual(sub.last_value, case["displayedAfterApply"])
        self.assertEqual(seen[-1], case["displayedAfterApply"])

        _frame(client, case["frame"])

        # The overlay survived the frame and was RE-FOLDED onto the new base,
        # rather than being clobbered by it.
        self.assertEqual(sub.last_value, case["displayedAfterFrame"])
        self.assertEqual(len(sub.optimistic_layers), case["layersAfterFrame"])
        self.assertEqual(seen[-1], case["displayedAfterFrame"])

    def test_a_throwing_layer_is_skipped_not_fatal(self):
        covers("optimistic_layer_rebases_onto_server_frame")
        case = FIXTURES["throwingLayerSkipped"]
        client, sub, seen = _live(case["base"])

        def explode(_current):
            raise RuntimeError("buggy optimistic update")

        # Registered directly rather than through apply_optimistic_layer, which
        # refuses a transform that raises on the value it is first handed. This
        # is the other case: a layer that worked once and raises on a later
        # rebase, which the fold must survive.
        sub.optimistic_layers.append(OptimisticLayer(explode))

        deferred = []
        apply_optimistic_layer(sub, _appender(case["appended"]), deferred)
        _run(deferred)

        _frame(client, {"data": case["base"]})

        self.assertEqual(len(sub.optimistic_layers), case["layers"])
        self.assertEqual(seen[-1], case["displayed"])


class TestOptimisticCommitCursor(unittest.TestCase):
    def test_layer_drops_only_once_a_frame_reaches_the_commit_cursor(self):
        covers("optimistic_layer_drops_on_commit_cursor")
        case = FIXTURES["commitCursorDrop"]
        client, sub, _seen = _live(case["base"])

        deferred = []
        handle = apply_optimistic_layer(sub, _appender(case["appended"]), deferred)
        handle.confirm(case["commitCursor"], deferred)
        _run(deferred)

        _frame(client, case["belowFrame"])

        # Below the commit cursor: the write is NOT in the server base yet, so
        # dropping the overlay here would blink the value away and back.
        self.assertEqual(sub.last_value, case["displayedAfterBelowFrame"])
        self.assertEqual(len(sub.optimistic_layers), case["layersAfterBelowFrame"])

        _frame(client, case["atFrame"])

        # The frame reached the commit cursor: the effect is in the base, so the
        # overlay drops without the value ever double-counting it.
        self.assertEqual(sub.last_value, case["displayedAfterAtFrame"])
        self.assertEqual(len(sub.optimistic_layers), case["layersAfterAtFrame"])

    def test_confirm_without_a_cursor_drops_without_reverting(self):
        covers("optimistic_layer_drops_on_commit_cursor")
        case = FIXTURES["confirmWithoutCursor"]
        _client, sub, _seen = _live(case["base"])

        deferred = []
        handle = apply_optimistic_layer(sub, _appender(case["appended"]), deferred)
        handle.confirm(None, deferred)
        _run(deferred)

        # CDC is off on this shard, so there is no cursor to gate on. The layer
        # goes, but the display does not revert: the write DID commit.
        self.assertEqual(sub.last_value, case["displayedAfterConfirm"])
        self.assertEqual(len(sub.optimistic_layers), case["layersAfterConfirm"])

    def test_a_confirmed_cursor_already_passed_drops_immediately(self):
        covers("optimistic_layer_drops_on_commit_cursor")
        case = FIXTURES["commitCursorDrop"]
        _client, sub, _seen = _live(case["atFrame"]["data"], cursor=case["atFrame"]["cursor"])

        deferred = []
        handle = apply_optimistic_layer(sub, _appender("x"), deferred)
        handle.confirm(case["commitCursor"], deferred)
        _run(deferred)

        # The confirming frame beat the RPC response — the common race. The
        # overlay must drop on confirm rather than linger until the next frame.
        self.assertEqual(sub.optimistic_layers, [])
        self.assertEqual(sub.last_value, case["atFrame"]["data"])

    def test_a_cursorless_frame_leaves_the_tracked_cursor_alone(self):
        covers("optimistic_cursorless_frame_preserves_cursor")
        case = FIXTURES["cursorlessFrame"]
        client, sub, seen = _live(case["base"])

        deferred = []
        handle = apply_optimistic_layer(sub, _appender(case["appended"]), deferred)
        _run(deferred)

        _frame(client, case["cursoredFrame"])
        _frame(client, case["cursorlessFrame"])

        # `cursor` is OPTIONAL on a data frame. Nulling the tracked cursor here
        # strands the layer: the confirm below compares against it, so the
        # overlay would survive its own commit and render the write twice.
        self.assertEqual(sub.server_cursor, case["cursorAfterCursorlessFrame"])
        self.assertEqual(sub.last_value, case["displayedAfterCursorlessFrame"])
        self.assertEqual(seen[-1], case["displayedAfterCursorlessFrame"])
        self.assertEqual(len(sub.optimistic_layers), case["layersAfterCursorlessFrame"])

        deferred = []
        handle.confirm(case["commitCursor"], deferred)
        _run(deferred)

        self.assertEqual(len(sub.optimistic_layers), case["layersAfterConfirm"])


class TestOptimisticRollback(unittest.TestCase):
    def test_rollback_restores_the_server_value(self):
        covers("optimistic_layer_rolls_back_on_failure")
        case = FIXTURES["rollback"]
        _client, sub, seen = _live(case["base"])

        deferred = []
        handle = apply_optimistic_layer(sub, _appender(case["appended"]), deferred)
        handle.rollback(deferred)
        _run(deferred)

        self.assertEqual(sub.last_value, case["displayedAfterRollback"])
        self.assertEqual(len(sub.optimistic_layers), case["layersAfterRollback"])
        self.assertEqual(seen[-1], case["displayedAfterRollback"])

    def test_a_constant_layer_masks_concurrent_server_changes(self):
        covers("optimistic_layer_rolls_back_on_failure")
        case = FIXTURES["constantMask"]
        client, sub, seen = _live(case["base"])
        client.detach_socket()

        # Driven through the real write path: the local store records the
        # override with the client's lock RELEASED and the client installs it.
        outcome = asyncio.run(
            client.submit(
                SubmitOptions(
                    args=ARGS,
                    function_path=PATH,
                    optimistic_update=lambda store, _args: store.set_query(PATH, ARGS, case["value"]),
                )
            )
        )

        self.assertTrue(outcome.queued)
        self.assertEqual(sub.last_value, case["displayedAfterApply"])

        _frame(client, case["frame"])

        # An absolute override: while pending it re-clamps and HIDES the
        # concurrent server change rather than merging with it.
        self.assertEqual(sub.last_value, case["displayedAfterFrame"])

        # Closing hands every queued write back, which rolls its layers off.
        client.close()

        self.assertEqual(sub.last_value, case["displayedAfterRollback"])
        self.assertEqual(seen[-1], case["displayedAfterRollback"])

    def test_the_store_reads_back_what_this_batch_wrote(self):
        covers("optimistic_layer_rolls_back_on_failure")
        _client, sub, _seen = _live(["a"])
        sub.args = ARGS
        store = OptimisticLocalStore(lambda _path, _args: [sub], lambda _path: [sub])

        self.assertEqual(store.get_all_queries(PATH), [(ARGS, ["a"])])

        store.set_query(PATH, ARGS, ["z"])

        # Recorded, not installed — but a second read in the same callback must
        # still see it, or a multi-step update composes onto a stale value.
        self.assertEqual(store.get_query(PATH, ARGS), ["z"])
        self.assertEqual(store.get_all_queries(PATH), [(ARGS, ["z"])])
        self.assertEqual(sub.last_value, ["a"])


if __name__ == "__main__":  # pragma: no cover
    unittest.main()

File: sdks/ruby/test/test_optimistic.rb
# frozen_string_literal: true

# The cursor-gated optimistic-layer engine, against the shared golden scenarios
# in protocol/fixtures/offline-optimistic.json. Every expectation is read from
# that file so this port and the other six assert the same values rather than
# each documenting its own behaviour.

require "json"
require "minitest/autorun"

require_relative "../lib/lunora"
require_relative "fixtures"
require_relative "manifest"

module OptimisticFixtures
  include FixtureLoader

  SUBSCRIPTION_ID = "sub_1"
  QUERY = "messages:list"

  def optimistic_case(name) = scenario("optimistic", name)

  # The one transform primitive the fixtures use: push onto a COPY of the list.
  #
  # A copy, not an in-place push: a transform is re-run on every rebase, so one
  # that mutated its input would compound its own effect on each server frame.
  def appender(item) = ->(current) { (current || []) + [item] }

  def build_state(base, seen = nil)
    Lunora::Optimistic::State.build(base, seen.nil? ? [] : [->(value) { seen << value }])
  end

  # A client with one live subscription primed to +base+, plus the list every
  # value delivered to that subscription lands in.
  #
  # The frames below go through +Client#handle_frame+ — the PRODUCTION path. The
  # suite used to drive a hand-copied transcription of its `data` branch instead,
  # which is why a `handleFrame` that forgot to advance the cursor, or to drop
  # confirmed layers, would have kept all nine conformance names green.
  def primed_client(base, cursor: 1, responses: nil)
    seen = []
    poster = responses.nil? ? nil : ->(_url, _headers, _body) { [200, responses] }
    client = Lunora::Client.new("https://app.example", http_post: poster)

    client.attach_socket(->(_frame) {})
    client.subscribe(QUERY, {}, ->(value) { seen << value })
    deliver(client, { "cursor" => cursor, "data" => base })

    [client, seen]
  end

  def deliver(client, frame, kind: "data")
    client.handle_frame(JSON.generate(frame.merge("id" => SUBSCRIPTION_ID, "type" => kind)))
  end

  # The subscription's tracked cursor and live layer count. Read through the
  # registry because neither is public API — and neither should be; they are
  # internal state the fixture pins so a rebase cannot quietly stop rebasing.
  def tracked(client)
    client.instance_variable_get(:@subscriptions).fetch(SUBSCRIPTION_ID)
  end
end

class TestOptimisticRebase < Minitest::Test
  include OptimisticFixtures

  def test_layer_rebases_onto_a_later_server_frame
    ConformanceManifest.covers("optimistic_layer_rebases_onto_server_frame")
    case_data = optimistic_case("rebase")
    client, seen = primed_client(case_data["base"])
    # Offline, so the write QUEUES and its overlay stays pending across the frame.
    client.detach_socket

    client.submit(QUERY, {}, optimistic: appender(case_data["appended"]))

    assert_equal case_data["displayedAfterApply"], seen.last

    deliver(client, case_data["frame"])

    # The overlay survived the frame and was RE-FOLDED onto the new base, rather
    # than being clobbered by it.
    assert_equal case_data["displayedAfterFrame"], seen.last
    assert_equal case_data["layersAfterFrame"], tracked(client)[:state].layers.length
  end

  # `cursor` is OPTIONAL on a data frame, and one that omits it must leave the
  # tracked cursor where it was: the cursor is what a later commitCursor is
  # compared against, so nulling it strands every pending layer and the write
  # renders twice until some later cursored frame happens to land.
  def test_a_cursorless_frame_leaves_the_tracked_cursor_alone
    ConformanceManifest.covers("optimistic_cursorless_frame_preserves_cursor")
    case_data = optimistic_case("cursorlessFrame")
    client, seen = primed_client(case_data["base"], responses: { "commitCursor" => case_data["commitCursor"], "result" => nil })
    client.detach_socket

    client.submit(QUERY, {}, optimistic: appender(case_data["appended"]))
    deliver(client, case_data["cursoredFrame"])
    deliver(client, case_data["cursorlessFrame"])

    assert_equal case_data["cursorAfterCursorlessFrame"], tracked(client)[:cursor]
    assert_equal case_data["displayedAfterCursorlessFrame"], seen.last
    assert_equal case_data["layersAfterCursorlessFrame"], tracked(client)[:state].layers.length

    client.attach_socket(->(_frame) {})
    client.flush_offline_queue

    # Only reachable because the cursor survived: confirm compares the write's
    # commit cursor against it.
    assert_equal case_data["layersAfterConfirm"], tracked(client)[:state].layers.length
  end

  def test_a_raising_layer_is_skipped_not_fatal
    ConformanceManifest.covers("optimistic_layer_rebases_onto_server_frame")
    case_data = optimistic_case("throwingLayerSkipped")
    state = build_state(case_data["base"])

    # Registered directly rather than through apply_layer, which refuses a
    # transform that raises on the value it is first handed. This is the other
    # case: a layer that worked once and raises on a later rebase, which the fold
    # must survive.
    state.layers << Lunora::Optimistic::Layer.new(0, ->(_current) { raise "buggy optimistic update" }, nil)

    deferred = []
    Lunora::Optimistic.apply_layer(state, appender(case_data["appended"]), deferred)
    deferred.each(&:call)

    assert_equal case_data["layers"], state.layers.length
    assert_equal case_data["displayed"], Lunora::Optimistic.fold(state.server_base, state.layers)
  end
end

class TestOptimisticCommitCursor < Minitest::Test
  include OptimisticFixtures

  def test_layer_drops_only_once_a_frame_reaches_the_commit_cursor
    ConformanceManifest.covers("optimistic_layer_drops_on_commit_cursor")
    case_data = optimistic_case("commitCursorDrop")
    client, seen = primed_client(case_data["base"], responses: { "commitCursor" => case_data["commitCursor"], "result" => nil })

    client.submit(QUERY, {}, optimistic: appender(case_data["appended"]))
    deliver(client, case_data["belowFrame"])

    # Below the commit cursor: the write is NOT in the server base yet, so
    # dropping the overlay here would blink the value away and back.
    assert_equal case_data["displayedAfterBelowFrame"], seen.last
    assert_equal case_data["layersAfterBelowFrame"], tracked(client)[:state].layers.length

    deliver(client, case_data["atFrame"])

    # The frame reached the commit cursor: the effect is in the base, so the
    # overlay drops without the value ever double-counting it.
    assert_equal case_data["displayedAfterAtFrame"], seen.last
    assert_equal case_data["layersAfterAtFrame"], tracked(client)[:state].layers.length
  end

  def test_layer_drops_on_a_settled_frame_reaching_the_commit_cursor
    ConformanceManifest.covers("optimistic_layer_drops_on_settled_frame")
    case_data = optimistic_case("settledFrameDrop")
    client, seen = primed_client(case_data["base"], responses: { "commitCursor" => case_data["commitCursor"], "result" => nil })

    client.submit(QUERY, {}, optimistic: appender(case_data["appended"]))
    deliver(client, case_data["belowFrame"], kind: "settled")

    assert_equal case_data["displayedAfterBelowFrame"], seen.last
    assert_equal case_data["layersAfterBelowFrame"], tracked(client)[:state].layers.length

    deliver(client, case_data["atFrame"], kind: "settled")

    # A byte-identical write yields a settled frame, never a data frame. Sweeping
    # only on data frames leaves the prediction on screen until some unrelated
    # write happens to change this query.
    assert_equal case_data["displayedAfterAtFrame"], seen.last
    assert_equal case_data["layersAfterAtFrame"], tracked(client)[:state].layers.length
  end

  def test_confirm_without_a_cursor_drops_without_reverting
    ConformanceManifest.covers("optimistic_layer_drops_on_commit_cursor")
    case_data = optimistic_case("confirmWithoutCursor")
    state = build_state(case_data["base"])

    deferred = []
    handle = Lunora::Optimistic.apply_layer(state, appender(case_data["appended"]), deferred)
    handle.confirm(nil, deferred)
    deferred.each(&:call)

    # CDC is off on this shard, so there is no cursor to gate on. The layer goes,
    # but the display does not revert: the write DID commit.
    assert_equal case_data["displayedAfterConfirm"], state.last_value
    assert_equal case_data["layersAfterConfirm"], state.layers.length
  end

  def test_a_confirmed_cursor_already_passed_drops_immediately
    ConformanceManifest.covers("optimistic_layer_drops_on_commit_cursor")
    case_data = optimistic_case("commitCursorDrop")
    state = build_state(case_data["atFrame"]["data"])
    state.server_cursor = case_data["atFrame"]["cursor"]

    deferred = []
    handle = Lunora::Optimistic.apply_layer(state, appender("x"), deferred)
    handle.confirm(case_data["commitCursor"], deferred)
    deferred.each(&:call)

    # The confirming frame beat the RPC response — the common race. The overlay
    # must drop on confirm rather than linger until the next frame.
    assert_empty state.layers
    assert_equal case_data["atFrame"]["data"], state.last_value
  end
end

class TestOptimisticRollback < Minitest::Test
  include OptimisticFixtures

  def test_rollback_restores_the_server_value
    ConformanceManifest.covers("optimistic_layer_rolls_back_on_failure")
    case_data = optimistic_case("rollback")
    seen = []
    state = build_state(case_data["base"], seen)

    deferred = []
    handle = Lunora::Optimistic.apply_layer(state, appender(case_data["appended"]), deferred)
    handle.rollback(deferred)
    deferred.each(&:call)

    assert_equal case_data["displayedAfterRollback"], state.last_value
    assert_equal case_data["layersAfterRollback"], state.layers.length
    assert_equal case_data["displayedAfterRollback"], seen.last
  end

  def test_a_constant_layer_masks_concurrent_server_changes
    ConformanceManifest.covers("optimistic_layer_rolls_back_on_failure")
    case_data = optimistic_case("constantMask")
    client, seen = primed_client(case_data["base"])
    client.detach_socket
    read_back = nil
    read_back_all = nil

    client.submit("messages:send", {}, optimistic_update: lambda { |store, _args|
      store.set_query(QUERY, {}, case_data["value"])
      # Reflects the override already written in this batch, before anything is
      # installed on the subscription — through BOTH readers, or a multi-step
      # update that patches every variant of a list and then reads it back
      # composes onto the value it just replaced.
      read_back = store.get_query(QUERY, {})
      read_back_all = store.get_all_queries(QUERY).map(&:last)
    })

    assert_equal case_data["displayedAfterApply"], seen.last
    assert_equal case_data["displayedAfterApply"], read_back
    assert_equal [case_data["displayedAfterApply"]], read_back_all

    deliver(client, case_data["frame"])

    # An absolute override: while pending it re-clamps and HIDES the concurrent
    # server change rather than merging with it.
    assert_equal case_data["displayedAfterFrame"], seen.last

    # Closing rejects the still-queued write, which unwinds its layers.
    client.close

    assert_equal case_data["displayedAfterRollback"], seen.last
  end
end
